src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -0,0 +1,349 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal interface ITypeImportCompletionService : IWorkspaceService
+    {
+        void GetAccessibleTopLevelTypesFromPEReference(
+            Solution solution,
+            Compilation compilation,
+            PortableExecutableReference peReference,
+            Action<CompletionItem> handleAccessibleItem,
+            CancellationToken cancellationToken);
+
+        Task GetAccessibleTopLevelTypesFromCompilationReferenceAsync(
+            Solution solution,
+            Compilation compilation,
+            CompilationReference compilationReference,
+            Action<CompletionItem> handleAccessibleItem,
+            CancellationToken cancellationToken);
```

Comment:
this method feels unnecessary.... it feels like the caller can just map the compilationreference to an appropriate project, and then call the method below, right?

Reply:
Actually no. The service is still handling IVTs right now, so it needs the this unless to push the responsibility to provider as well.

Reply:
i would be ok with that.  i like the idea of hte service simply caching the items for source and metadata-symbols.

Reply:
Done